### DISCUSSION TITLE
Support for importing recipes from meal-kit services

### DISCUSSION THREAD

**Discussion by aureateflux** [mealie-recipes/mealie#2391] [Feature Request]
**Before submitting this feature request I have**
_COMPLETE AND FILL THESE OUT_
- [x] Reviewed the documentation and verified that my feature request isn't already a feature.
- [x] Reviewed the existing feature requests and my feature has not been requested.
- [x] Checked the [tasks tagged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

**Please Describe The Problem To Be Solved**
Mealie works great for scraping recipes from recipe blogs, but if you want to add recipes from a meal-kit service to your meal planning, the work flow is very rough. URL import doesn't work and manual import requires converting PDF to PNG and adding steps and ingredients manually.

There are two problems with importing recipes from these services: 1) Services generally gate the recipes behind a sign-in; 2) They don't follow the recipe mark-up conventions that most recipe blog websites follow.

**(Optional): Suggest A Solution**
Solution 1 (Directly scrape from meal-kit website):
While they don't follow the blog recipe conventions, meal-kit services do tend to structure their recipes similarly to one another, and it should be possible to set up an import scheme that would be successful with most of them.

Mealie would then need support for gaining access to sign-in gated recipes, either through access token or user-provided credentials so the user can provide a URL (or it may be necessary to add some browser functionality to acount for differences in website functionality from service to service).

Solution 2 (PDF upload support)
Many meal-kit services provide the option of downloading PDFs of the recipes, and those that don't provide paper recipes that could be scanned and uploaded manually. There's already an old feature request for PDF import, but there may still need to be work done to automate more of the import process specifically for meal-kit recipes (or just improve the UX for manual imports in general).

Solution 1 would provide a better user experience overall, especially if it reaches parity with the recipe scraper support for blogs. PDF support may be quicker to implement and would add a valuable and versatile feature.

**Additional Information**

- [ ] If the feature is accepted I'm willing to submit a PR to provide this feature

**Comment by michael-genson** [comment]
Mealie already supports rich migrations from other recipe management services via data migrations, separate from URL scraping. Currently there's support for Nextcloud, Chowdown, Copy Me That, and Paprika. If there's another app you'd like to see supported, let us know and we can take a look!

Separately, Mealie also has OCR support, so you can import recipes by scan using the built-in OCR tool to extract text without needing to copy everything manually.

**Reply by aureateflux** [reply]
Using OCR to import PDF recipes involves first converting the PDF to PNGs, then uploading PNGs for OCR. Like I said, if you could import PDFs it becomes largely a UX problem.

When I tried it, I had a hard time getting the OCR editing tool to put selected text into the Step fields without doing a bulk add, but with bulk add once the fields were populated I couldn't figure out how to edit anything to fix OCR errors or combine steps. This was either a me problem, a UX problem, or both, and as I understand it is already being worked on.

As for the import, I'm specifically talking about recipes provided by meal-kit services like HelloFresh, Dinnerly, Home Chef, etc.

**Reply by michael-genson** [reply]
Ah I see, "meal kits" didn't click for me, but I'm familiar with those examples
I agree that the OCR import could use additional work, particularly to work with PDFs

**Comment by jack5mikemotown** [comment]
I want to add that this would be great too, i get a lot of my recipes from meal kit sites. While a few meal kit sites do support scraping without a login, the vast majority do not. Most offer a PDF version of their recipes which is easy enough and while it does require a bit more work than scraping, still does a lot of the work for you. It would be great though if an option for using OCR on PDFs was available as that would at least be a nice easy solution to get the recipe started without having to throw the PDF through another app

**Comment by hay-kot** [comment]
I think the generally better solution here is to implement a HTML api where you can provide HTML to be scraped which can be passed to the recipe scraping library. I had submitted a PR upstream to support HTML directly with this intention in mind but never got around to implementing it. 

Accepting HTML directly would 1) allow us to build an Upload HTML section in the UI and 2) allow someone to build a chrome extension that would do that whole process for them. 


> (PDF upload support)
Many meal-kit services provide the option of downloading PDFs of the recipes, and those that don't provide paper recipes that could be scanned and uploaded manually. There's already an old feature request for PDF import, but there may still need to be work done to automate more of the import process specifically for meal-kit recipes (or just improve the UX for manual imports in general).

I genuinely have no idea how this could be implement in a generic way. Working with PDFs reliably is basically impossible.

**Reply by michael-genson** [reply]
> Working with PDFs reliably is basically impossible

The only way I can see this working consistently is to treat it as an image and use OCR

**Reply by jack5mikemotown** [reply]
I could be totally off base here because I don't know much about what it takes to do any of this, but I think treating a PDF as an image for OCR purposes is a useful function to have. My thoughts with this one in particular are that it's probably not particularly difficult to implement and PDFs are widely used, at least in the meal kit space, for downloadable recipes, so at least that provides some sort of semi-automated way to grab recipes that come as pdfs. An HTML solution definitely makes more sense in general, but I think support for PDFs as images would be a nice fallback

**Reply by aureateflux** [reply]
> The only way I can see this working consistently is to treat it as an image and use OCR

This is basically what Paperless does, except it converts it back to a PDF with the OCR data attached. From that approach, the only thing Mealie is really missing is converting the PDF to PNG images (and adding handling for multi-page PDFs). OCR of PNGs is already in place with PNG import and there's no need to store an OCR'd PDF.

**Comment by michael-genson** [comment]
FYI you can now import recipes manually using HTML or JSON: https://github.com/mealie-recipes/mealie/pull/4274 (the JSON has to be in https://schema.org/Recipe format)

If we can't reach the site (e.g. meal-kit services that require a login) you can manually copy/paste the HTML from the site yourself

For the use case of PDFs, your best bet is still converting to an image and importing that way (which uses OpenAI for OCR)
